### PR TITLE
pysisyphus: init at 0.7rc1

### PR DESCRIPTION
**Pull request by sheepforce**
This adds Pysisyphus for geometry optimisaiton methods. It implements optimisations with wavefunction root tracking, chain of state methods, AFIR and more methods for transition, ground and excited state optimisations. Pysisyphus wraps many other QC programs to obtain forces and is somehow difficult to setup without nix.

As PySCF currently fails to build for my, this PR is blocked by #29 .

This and the GAMESS-US PR #28 are the last package additions from my list. :slightly_smiling_face: :partying_face:

### CODE REVIEW

File: pysisyphus/default.nix
Comment (markuskowa) [inline pysisyphus/default.nix:27]: This should be sufficient.
```suggestion
        cmd = "${psi4Wrapper}";
```

File: pysisyphus/default.nix
Comment (markuskowa) [inline pysisyphus/default.nix:65]:
```suggestion
    propagatedBuildInputs = with lib; [
```

**Review by markuskowa — CHANGES_REQUESTED** [review]
pysisyphus itself builds fine for me. However, the test does not build:
```
# (in test file /nix/store/50na6a4b1q59xs0z92f9kwimiw82pk1f-psi4.bats, line 21)
#   `/nix/store/pryg0kagvqgl5bi0p3vrq413k1pjwgk1-python3.8-pysisyphus-0.6/bin/pysis pysis.yaml' failed
# [1619822613.273416] [localhost:49   :0]      tcp_iface.c:602  UCX  ERROR opendir(/sys/class/net) failed: No such file or directory
# /nix/store/nvvrs0sczfwhl45gmwwixzqj7m8x8rpx-python3.8-pyscf-1.7.3/lib/python3.8/site-packages/pyscf/lib/misc.py:47: H5pyDeprecationWarning: Using default_file_mode other than 'r' is deprecated. Pass the mode to h5py.File() instead.
#   h5py.get_config().default_file_mode = 'a'
# Read pysisyphus configuration from '/nix/store/pryg0kagvqgl5bi0p3vrq413k1pjwgk1-python3.8-pysisyphus-0.6/share/pysisyphus/pysisrc'
# xtb-python is not available!
# Traceback (most recent call last):
#   File "/nix/store/pryg0kagvqgl5bi0p3vrq413k1pjwgk1-python3.8-pysisyphus-0.6/bin/..pysis-wrapped-wrapped", line 6, in <module>
#     from pysisyphus.run import run
#   File "/nix/store/pryg0kagvqgl5bi0p3vrq413k1pjwgk1-python3.8-pysisyphus-0.6/lib/python3.8/site-packages/pysisyphus/run.py", line 25, in <module>
#     from pysisyphus.calculators import *
#   File "/nix/store/pryg0kagvqgl5bi0p3vrq413k1pjwgk1-python3.8-pysisyphus-0.6/lib/python3.8/site-packages/pysisyphus/calculators/__init__.py", line 38, in <module>
#     from pysisyphus.calculators.ONIOMv2 import ONIOM
#   File "/nix/store/pryg0kagvqgl5bi0p3vrq413k1pjwgk1-python3.8-pysisyphus-0.6/lib/python3.8/site-packages/pysisyphus/calculators/ONIOMv2.py", line 48, in <module>
#     from pysisyphus.calculators.PySCF import PySCF
#   File "/nix/store/pryg0kagvqgl5bi0p3vrq413k1pjwgk1-python3.8-pysisyphus-0.6/lib/python3.8/site-packages/pysisyphus/calculators/PySCF.py", line 7, in <module>
#     from pyscf.dft import xcfun
#   File "/nix/store/nvvrs0sczfwhl45gmwwixzqj7m8x8rpx-python3.8-pyscf-1.7.3/lib/python3.8/site-packages/pyscf/dft/xcfun.py", line 32, in <module>
#     _itrf = lib.load_library('libxcfun_itrf')
#   File "/nix/store/nvvrs0sczfwhl45gmwwixzqj7m8x8rpx-python3.8-pyscf-1.7.3/lib/python3.8/site-packages/pyscf/lib/misc.py", line 69, in load_library
#     return numpy.ctypeslib.load_library(libname, _loaderpath)
#   File "/nix/store/cx0py33vl5li9yinz67rbky8g0lrv3wf-python3.8-numpy-1.20.2/lib/python3.8/site-packages/numpy/ctypeslib.py", line 153, in load_library
#     raise OSError("no file with expected extension")
# OSError: no file with expected extension
builder for '/nix/store/rpvm5fnivwvpmf25jmn2y4zyc5xzmmdk-psi4.drv' failed with exit code 1
```

**Comment by sheepforce** [comment]
At the moment I cannot reproduce the error, as for me the PySCF dependency already fails. I've tried with disabled PySCF tests and still get the AM1 errors and H5Py deprecation warnings when trying to run Pysisyphus. I will have a look as soon as PySCF is working.

**Comment by sheepforce** [comment]
So this now makes pysisyphus working :slightly_smiling_face: A minor PySCF update was necessary due to [Pyscf/#872](https://github.com/pyscf/pyscf/issues/872). I've also added the XTB-python library for more efficient interfacing.

The tests from the check phase do not work within nix right now and I guess this is due to how they are written.

@markuskowa I've added bats test. The Molcas and Turbomole ones do not work in the bats test, while they work fine in a nix-shell. The actually throw funny python errors. Do you have an idea what causes trouble here?

@eljost Most of the assertions do not pass. XTB needs more cycles and converges to a higher energy, The Molcas does not converge and does not look as if it ever would and the Turbomole test raises a `KeyError 'xyz'`. Could you please have a look what's going on there?

**Comment by markuskowa** [comment]
Under which conditions is the molcas test supposed to pass?
The molcas test also fails in Hydra:
```
#                          -----------------------------
#                          | HESSIAN AT FINAL GEOMETRY |
#                          -----------------------------
#
# ... mass-weighing cartesian hessian
# ... doing Eckart-projection
#
# First 10 eigenvalues [-1.4938e-02 -2.2596e-17 -4.0637e-18 -1.2579e-18  5.9134e-19  1.3083e-17
#   2.1119e-17  1.0916e-05  2.4721e-05  2.6807e-05]
# Imaginary frequencies: [-627.83] cm⁻¹
#
# Wrote final (not mass-weighted) hessian to 'calculated_final_cart_hessian'.
# Wrote HD5 Hessian to 'final_hessian.h5'.
# Wrote imaginary mode with ṽ=-627.83 cm⁻¹ to 'imaginary_mode_000.trj'
#
# Barrier between TS and first COS image: 55.8 kJ mol⁻¹
# Barrier between TS and last COS image: 138.1 kJ mol⁻¹
#
#
#                              #####################
#                              # ASSERTING RESULTS #
#                              #####################
# 00: opt.cur_cycle
#       Reference: 24
#         Current: 24
#         Matches: True
# 01: ts_opt.is_converged
#       Reference: True
#         Current: True
#         Matches: True
# 02: ts_opt.cur_cycle
#       Reference: 10
#         Current: 15
#         Matches: False
# 03: ts_geom._energy
#       Reference: -80.70048732
#         Current: -80.680988069325
#         Matches: False
# Traceback (most recent call last):
#   File "/nix/store/01cljm6c2jshiksslk3966m7sgf2nyzs-python3.8-pysisyphus-0.6/bin/..pysis-wrapped-wrapped", line 9, in <module>
#     sys.exit(run())
#   File "/nix/store/01cljm6c2jshiksslk3966m7sgf2nyzs-python3.8-pysisyphus-0.6/lib/python3.8/site-packages/pysisyphus/run.py", line 1946, in run
#     return run_from_dict(run_dict, **run_kwargs)
#   File "/nix/store/01cljm6c2jshiksslk3966m7sgf2nyzs-python3.8-pysisyphus-0.6/lib/python3.8/site-packages/pysisyphus/run.py", line 1902, in run_from_dict
#     check_asserts(run_result, run_dict)
#   File "/nix/store/01cljm6c2jshiksslk3966m7sgf2nyzs-python3.8-pysisyphus-0.6/lib/python3.8/site-packages/pysisyphus/run.py", line 1633, in check_asserts
#     assert all(matches)
# AssertionError
builder for '/nix/store/pxl50w4kk83prk0035vkag46imp4ssa6-psi4.drv' failed with exit code 1
```

**Comment by eljost** [comment]
> @eljost Most of the assertions do not pass. XTB needs more cycles and converges to a higher energy, The Molcas does not converge and does not look as if it ever would and the Turbomole test raises a `KeyError 'xyz'`. Could you please have a look what's going on there?

Which tests do not pass? The files under `examples` are not part of my CI test suite, so maybe not all of the examples there are up to date...

**Comment by sheepforce** [comment]
@eljost Here is a quick list of things from the examples that fail with outputs:
- [01_turbomole_cytosin_s1_opt](https://github.com/markuskowa/NixOS-QChem/files/6521199/01_turbomole_cytosin_s1_opt.out.txt) `KeyError: 'xyz'`
- [03_xtb_neb_optimization](https://github.com/markuskowa/NixOS-QChem/files/6521242/03_xtb_neb_optimization.out.txt) more cycles, higher energy and wild XTB timings
- [07_openmolcas_trans_butadiene_s1_opt](https://github.com/markuskowa/NixOS-QChem/files/6521315/07_openmolcas_trans_butadiene_s1_opt.out.txt) [optimizer.log](https://github.com/markuskowa/NixOS-QChem/files/6521323/optimizer.log) (incomplete) doesn't look as if it would go anywhere in the optimisation. Is quite far of still in the energies from what the assertion expects

@markuskowa I guess the tests are always supposed to pass, but let's wait for what @eljost can tell us there. Regarding the Molcas example (you got the XTB example); it runs in a `nix-shell` with the output above, but it produces a completely different error in the bats test: [molcas_bats.txt](https://github.com/markuskowa/NixOS-QChem/files/6521337/molcas_bats.txt)

**Comment by markuskowa** [comment]
> but it produces a completely different error in the bats test: molcas_bats.txt.

The snippet above is taken from the hydra build of `qchem.tests.pysisyphus`.

> I guess the UCX error right at the start could be the problem? Although I have no idea why this occurs.

This is a sandbox issue, since there is no `/sys` mounted. I have seen this in other builds, but I do not think it is a problem here.

**Comment by sheepforce** [comment]
> The snippet above is taken from the hydra build of `qchem.tests.pysisyphus`.

Ah yes. It's the XTB part that fails the assertion error in your previous post. This has been fixed now. But also the Molcas test failed with an error that only happens in the bats test.

However, I sat down with @eljost and we had a look at the test suite. The assumption that they are only executed from within a Git repo has been fixed. The pytest now actually runs properly with Nix. At the moment we are pinning a commit from a hotfix branch, but before merging we can wait for a new release from @eljost with the fixes included.

@eljost The full test suite still has some failures. Can you have a look? You can execute/build/test like this `nix-build`
[`pysisTest.nix`](https://github.com/markuskowa/NixOS-QChem/files/6523399/pysisTest.nix.txt)`

`

File: pkgs/apps/nwchem/default.nix
Comment (markuskowa) [inline pkgs/apps/nwchem/default.nix:6]: true ;)

File: pkgs/apps/pysisyphus/default.nix
Comment (markuskowa) [inline pkgs/apps/pysisyphus/default.nix:47]:
```suggestion
        { name = "openmolcas"; value.cmd = "${openmolcas}/bin/pymolcas"; }
```
This pattern would save a lot of space here.

File: pkgs/apps/xtb/default.nix
Comment (markuskowa) [inline pkgs/apps/xtb/default.nix:1]:
```suggestion
{ stdenv, lib, gfortran, fetchFromGitHub, cmake, makeWrapper, writeTextFile, mkl
, ...
} :
```
nixpkgs style

File: pkgs/apps/xtb/default.nix
Comment (markuskowa) [inline pkgs/apps/xtb/default.nix:51]: Instead of creating a file in the nix-store first, I would write the file directly in the `postInstall` phase:
```
cat > $out/lib/pkgconfig/xtb.pc << EOF
...
EOF
```
Here you have shell substitution of `$out` automatically.

File: pkgs/apps/xtb/default.nix
Comment (markuskowa) [inline pkgs/apps/xtb/default.nix:56]: `binSearchPath` does not need to available in the derivation?
better:
```
postFixup = let 
  binSearchPath = ...;
in ''
''
wrapProgram ....
'';

File: pkgs/lib/rmsd/default.nix
Comment (markuskowa) [inline pkgs/lib/rmsd/default.nix:11]: `meta` = ?

File: pythonPackages.nix
Comment (markuskowa) [inline pythonPackages.nix:56]: `molpro` is now `null` if no license key is provided. We can drop the conditionals checking for `cfg.licMolpro`.

**Comment by sheepforce** [comment]
@markuskowa All good suggestions :slightly_smiling_face: I will take care of those when Pysis actually works. More and more strange behaviour is popping up here. The Turbomole tests only pass if CFour is present and stuff like this ... 

I will probably meet with @eljost next week and we will try to get to solve the remaining issues and then I will finalise the PR.

File: pythonPackages.nix
Comment (sheepforce) [reply pythonPackages.nix]: fixed for molpro but `gaussian` is still an unknown attribute to `selfPkgs`. This is simply missing if `optpath` is not set, right?

File: pkgs/apps/pysisyphus/default.nix
Comment (markuskowa) [inline pkgs/apps/pysisyphus/default.nix:15]:
```suggestion
    #!${runtimeShell}
```

File: pkgs/apps/pysisyphus/default.nix
Comment (markuskowa) [inline pkgs/apps/pysisyphus/default.nix:105]: That is better placed in the top level `let`.

File: pkgs/apps/qcelemental/default.nix
Comment (markuskowa) [inline pkgs/apps/qcelemental/default.nix:19]: When the upstream fix is merged we can remove `qcelemental` and the patches.

Comment (sheepforce) [reply pkgs/apps/qcelemental/default.nix]: It is not in nixpkgs-unstable yet so I've kept it for the moment in the same version as on nixpkgs master?

File: pythonPackages.nix
Comment (sheepforce) [reply pythonPackages.nix]: I've did the same to gaussian as you did to molpro as a workaround :slightly_smiling_face:

**Comment by sheepforce** [comment]
OK, thanks to @eljost we could figure out a lot of problems, including the QCElemental bug and now it's running just fine. From my perspective we are good to go. :slightly_smiling_face:

File: pkgs/apps/qcelemental/default.nix
Comment (markuskowa) [reply pkgs/apps/qcelemental/default.nix]: Yes, that is the disadvantage of `nixpkgs-unstable` here. It can take a few days while until the channels advances.

**Review by markuskowa — APPROVED** [review]

**Pull request merged**
